### PR TITLE
Circuit-break based on real memory usage

### PR DESCRIPTION
**Pull request by danielmitterdorfer**
With this commit we introduce a new circuit-breaking strategy to the parent
circuit breaker. Contrary to the current implementation which only accounts for
memory reserved via child circuit breakers, the new strategy measures real heap
memory usage at the time of reservation. This allows us to be much more
aggressive with the circuit breaker limit so we bump it to 95% by default. The
new strategy is turned on by default and can be controlled  with the new cluster
setting `indices.breaker.total.use_real_memory`.

Note that we turn it off for all integration tests with an internal test cluster
because it leads to spurious test failures which are of no value (we cannot
fully control heap memory usage in tests). All REST tests, however, will make
use of the real memory circuit breaker.

### CODE REVIEW

**Comment by danielmitterdorfer** [comment]
A few more comments on this change for reviewers:

Initially I was worried about the overhead of `memoryMXBean.getHeapMemoryUsage().getUsed()` for every check in the parent breaker. Therefore I created a microbenchmark (see `MemoryStatsBenchmark`) and ran it on several machines and different JDK versions:

| CPU                                                          | JDK           | Duration per call - 1 thread [ns] | Duration per call - 64 threads [ns] | 
|--------------------------------------------------------------|---------------|-----------------------------------|-------------------------------------|
| Intel(R) Xeon(R) CPU E3-1246 v3 @ 3.50GHz (8 cores with HT)  | 1.8.0_131-b11 | 358 ± 11                          | --                                    | 
| Intel(R) Xeon(R) CPU E5-2686 v4 @ 2.30GHz (64 cores with HT) | 10.0.1+10     | 405 ± 1                           | 885 ± 10                            | 

CPUs ran at their base frequency with the performance CPU governor to reduce measurement noise.

So we can expect an overhead of several hundred nanoseconds per call even if we call that operation from lots of threads concurrently. Therefore, I concluded that this operation is cheap enough to call for every check in the parent breaker.

I also tested the effectiveness of this new circuit breaker strategy with several macrobenchmarks using the Elasticsearch default distribution with X-Pack Security enabled. I set the heap size for Elasticsearch to 256MB and with the new circuit breaker strategy it could absorb the load in all but one of the benchmarks in our macrobenchmarking suite (bulk-indexing and querying). Of course, the clients got request errors from Elasticsearch - for the full-text benchmark (`pmc`) as high as 30% - because the circuit breaker did its job. The point of these benchmarks was to show that even if we put a lot of pressure on Elasticsearch it can sustain. In the single node case (one shard, zero replicas), Elasticsearch never had an `OutOfMemoryError`. In the three-node case (one shard, one replica) we had one node die in the `nyc_taxis` benchmark while bulk-indexing. Perhaps a bulk-size of 10.000 documents and 8 concurrent clients was still too much but we plan to attack this with follow-up PRs.

Finally, I did a dedicated test to see how Elasticsearch behaves when an aggregation produces a crazy amount of buckets on a larger heap (16GB). I ingested the whole `http_logs` corpus (31GB of source documents). Then I intentionally set `search.max_buckets: 1000000000` and ran the following aggregation:

```json
{
  "aggs": {
    "size_over_time": {
      "date_histogram": {
        "field": "@timestamp",
        "interval": "second"
      },
      "aggs": {
        "sizes": {
          "percentiles": {
            "field": "size",
            "tdigest": {
              "compression": 1000
            }
          }
        }
      }
    }
  }
}
```

| Configuration   | `allow_partial_search_results`=true | `allow_partial_search_results = false` | 
|-----------------|-------------------------------------|----------------------------------------| 
| With this PR    | response after 73 seconds           | response after 21 minutes              | 
| Without this PR | node went OOM after 22 minutes      | node went OOM after 28 minutes         | 


While this PR greatly improves the situation w.r.t `OutOfMemoryError`s in Elasticsearch, this approach is still not perfect. As we do not fully manage heap allocations ourselves (e.g. in a pool) but rather rely on Java's `new` operator, reservation of memory in the circuit breaker and the actual allocation do not occur atomically. This means that it is still possible that the circuit breaker allows a reservation but if the time window between reservation and allocation is large enough, it is still possible that `OutOfMemoryError`s happen.

**Review by dakrone — CHANGES_REQUESTED** [review]
I left some comments, excited for this to get in though!

File: server/src/main/java/org/elasticsearch/indices/breaker/HierarchyCircuitBreakerService.java
Comment (danielmitterdorfer) [reply server/src/main/java/org/elasticsearch/indices/breaker/HierarchyCircuitBreakerService.java]: Of course. That was definitely unintentional.

File: server/src/main/java/org/elasticsearch/indices/breaker/HierarchyCircuitBreakerService.java
Comment (danielmitterdorfer) [reply server/src/main/java/org/elasticsearch/indices/breaker/HierarchyCircuitBreakerService.java]: It returns the maximum configured heap, i.e. they would get `8589934592` (i.e. 8GB) in your example (I verified this).

File: server/src/main/java/org/elasticsearch/indices/breaker/HierarchyCircuitBreakerService.java
Comment (danielmitterdorfer) [reply server/src/main/java/org/elasticsearch/indices/breaker/HierarchyCircuitBreakerService.java]: No, because this is for the current strategy which sums up the total memory reserved by all child circuit breakers. As the corresponding child circuit breaker accounts for that amount of memory already, we do not need to do that again in the parent breaker.

For the new strategy which is based on real memory usage, we do not rely on the child memory circuit breakers but rather only on current memory usage. Hence, we need to consider the amount of memory that is about to be reserved here.

File: test/framework/src/main/java/org/elasticsearch/test/InternalTestCluster.java
Comment (danielmitterdorfer) [reply test/framework/src/main/java/org/elasticsearch/test/InternalTestCluster.java]: We do. I have the impression that this is mainly caused by two facts: 

1. We run multiple integration tests in the same JVM process so garbage will add up.
2. We do not set a garbage collection algorithm in the integration tests (except for our g1 builds via `-Dtests.jvm.argline` in the Jenkins configuration). This means we let the JVM choose the algorithm. For the Java 10 builds this is `G1`, for Java 8 it is `ParallelGC`. From my tests I had the impression that especially G1 had trouble keeping up with allocation in our tests (as G1 cleans up concurrently) and I experienced circuit breaker issues basically all the time. 

 I am not sure how we should move forward here in general. G1 is not really optimal for such small heaps but on the other hand we should also run our tests with G1. I already added this to our group's agenda to discuss. Finally, the 95% are heap usage + amount to be reserved.

File: docs/reference/modules/indices/circuit_breaker.asciidoc
Comment (danielmitterdorfer) [reply docs/reference/modules/indices/circuit_breaker.asciidoc]: Makes sense. I'll change it.

File: docs/reference/modules/indices/circuit_breaker.asciidoc
Comment (danielmitterdorfer) [reply docs/reference/modules/indices/circuit_breaker.asciidoc]: I am happy to change the default to the real memory circuit breaker for all cases but initially I wanted to be a bit more conservative. I felt that for larger deployments it is ok to stick with the current implementation. I also thought that it might be one thing less to consider in a major version upgrade if we keep the default for now for those deployments.

My idea behind changing it only for lower heap sizes below 1GB deviation between real memory usage and actively tracked memory usage starts to matter more and we want to ensure we push back accordingly. I did not choose this heap size by coincidence but based on our benchmarks. With 1GB heap, Elasticsearch can handle our macrobenchmark suite but it starts to struggle below that point (e.g. for 768MB).

I could ask how the rest of the team feels turning it on in all cases. Wdyt?

**Comment by danielmitterdorfer** [comment]
@elasticmachine test this please.

File: server/src/main/java/org/elasticsearch/indices/breaker/HierarchyCircuitBreakerService.java
Comment (dakrone) [reply server/src/main/java/org/elasticsearch/indices/breaker/HierarchyCircuitBreakerService.java]: :+1:

File: test/framework/src/main/java/org/elasticsearch/test/InternalTestCluster.java
Comment (dakrone) [reply test/framework/src/main/java/org/elasticsearch/test/InternalTestCluster.java]: > Finally, the 95% are heap usage + amount to be reserved.

I suppose in that case it would force a real GC if the heap were at 95% and then a new amount was reserved using `new`

**Review by dakrone — APPROVED** [review]
> My idea behind changing it only for lower heap sizes below 1GB deviation between real memory usage and actively tracked memory usage starts to matter more and we want to ensure we push back accordingly. I did not choose this heap size by coincidence but based on our benchmarks. With 1GB heap, Elasticsearch can handle our macrobenchmark suite but it starts to struggle below that point (e.g. for 768MB).

> I could ask how the rest of the team feels turning it on in all cases. Wdyt?

I think that would be good to pursue, but that doesn't need to block this PR, I think we can discuss it for a follow-up.

LGTM assuming CI is happy :)

File: test/framework/src/main/java/org/elasticsearch/test/InternalTestCluster.java
Comment (danielmitterdorfer) [reply test/framework/src/main/java/org/elasticsearch/test/InternalTestCluster.java]: Actually the GC should get triggered even earlier. For CMS we have configured `-XX:CMSInitiatingOccupancyFraction=75`. But as the garbage collector is running concurrently with the application, it can still allocate while the GC is active.

**Comment by danielmitterdorfer** [comment]
@dakrone after our discussion yesterday I pushed the following changes:

* As we cannot fully control heap memory usage in the JVM-internal cluster tests and their purpose is also not testing circuit breakers (that's done by the REST tests), I have disabled the real-memory circuit breaker there.
* I have enabled the real memory circuit breaker by default now for all heap sizes. Consequently, I marked the PR now as "breaking".

Can you please have one more look at these changes and give your ok if you're fine with them?

Also, in one of my (very many) test runs I spotted the following test error:

```
REPRODUCE WITH: ./gradlew :server:integTest -Dtests.seed=71A44F93AD4E637B -Dtests.class=org.elasticsearch.search.aggregations.metrics.HDRPercentileRanksIT -Dtests.method="testScriptSingleValuedWithParams" -Dtests.security.manager=true -Dtests.locale=zh -Dtests.timezone=Asia/Saigon
FAILURE 0.16s J3 | HDRPercentileRanksIT.testScriptSingleValuedWithParams <<< FAILURES!
   > Throwable #1: java.lang.AssertionError: Count is 8 but 10 was expected.  Total shards: 9 Successful shards: 8 & 1 shard failures:
   >  shard [[DcF7TxxQQe-8jwOROrfM5A][idx][8]], reason [RemoteTransportException[[node_s0][127.0.0.1:62368][indices:data/read/search[phase/query/id]]]; nested: IllegalArgumentException[committed = 542113792 should be < max = 536870912]; ], cause [java.lang.IllegalArgumentException: committed = 542113792 should be < max = 536870912
   >    at java.lang.management.MemoryUsage.<init>(MemoryUsage.java:166)
   >    at sun.management.MemoryImpl.getMemoryUsage0(Native Method)
   >    at sun.management.MemoryImpl.getHeapMemoryUsage(MemoryImpl.java:71)
   >    at org.elasticsearch.indices.breaker.HierarchyCircuitBreakerService.currentMemoryUsage(HierarchyCircuitBreakerService.java:246)
   >    at org.elasticsearch.indices.breaker.HierarchyCircuitBreakerService.parentUsed(HierarchyCircuitBreakerService.java:234)
   >    at org.elasticsearch.indices.breaker.HierarchyCircuitBreakerService.checkParentLimit(HierarchyCircuitBreakerService.java:253)
[...]
```

This happened on MacOS 10.12.6 with JDK 10 (build 10.0.1+10). The stack trace points to an issue in the (native code of the) JDK. After analyzing the code, I suspect that the problem is the calculation of `committed`, not `max`, because `max` matches the 512MB heap size that we specify for our tests to the byte but committed is 517MB. Searching for [related issues in the OpenJDK bug tracker](https://bugs.openjdk.java.net/browse/JDK-8020530?jql=project%20%3D%20JDK%20AND%20issuetype%20%3D%20Bug%20AND%20component%20%3D%20core-svc%20AND%20text%20~%20%22should%20be%20%3C%20max%22%20AND%20Subcomponent%20%3D%20java.lang.management) only reveal an issue with non-heap memory but not with heap-memory. I am not aware of any prior discussion on the OpenJDK mailing lists about this topic. However, there is evidence that this happened before as well:

* There is a [ticket in the Netbeans bug tracker](https://netbeans.org/bugzilla/show_bug.cgi?id=194733) that has been closed with "To the best of my knowledge, this is a JDK bug." (without any reference).
* Apache Ignite has [implemented a workaround](https://github.com/apache/ignite/blob/df4fd65a32dcbd3db8f86db4a320b649b7a9a7e7/modules/core/src/main/java/org/apache/ignite/internal/managers/discovery/GridDiscoveryManager.java#L336-L346).

I suggest that we handle this as follows:

* Provided you agree, I will merge this PR to master as is, i.e. I will intentionally merge it with the knowledge that we are now exposed to this JDK bug on master. The idea is that this will give us more CI coverage in the hope that we find conditions under which this is reproducible. We probably need to turn on JVM logging to get more info.
* I will follow-up on the OpenJDK mailing list asking for feedback how to proceed. I doubt that an issue in the OpenJDK repo will be raised at this point, simply because I could not reproduce it. My hope is that we get some feedback that will help us to get to the bottom of this together with the OpenJDK developers so we can finally raise a ticket against OpenJDK.
* In case we do not get to the bottom of this before 7.0.0 is out, we should implement a workaround along the lines of:

```java
    long currentMemoryUsage() {
        try {
            return MEMORY_MX_BEAN.getHeapMemoryUsage().getUsed();
        } catch (IllegalArgumentException ex) {
            logger.warn("Could not determine real memory usage due to JDK issue.", ex);
            return 0L;
        }
    }
```

I think it is better to avoid the `IllegalArgumentException` bubbling up in this case.

As discussed yesterday, this change will also be backported to 6.x but the real memory circuit breaker will be turned *off* by default (I'll create a separate PR for this). I suggest that we integrate this workaround already in the backport.

**Review by dakrone — APPROVED** [review]
This LGTM, the steps you outlined in regard to the JDK bug sound good as well.

I noticed the Netbeans bug report is for OSX also, so perhaps it's an OSX-only issue?

**Comment by danielmitterdorfer** [comment]
Thank you for the review @dakrone! The JDK issue that I mentioned above is now raised against the OpenJDK repo in https://bugs.openjdk.java.net/browse/JDK-8207200.

**Comment by Bukhtawar** [comment]
When are we planning on back porting this to 6.x versions even now that the OpenJDK [bug](https://bugs.openjdk.java.net/browse/JDK-8207200) stands resolved

**Comment by yakirgb** [comment]
Hi, we are getting error from breaker about "Data too large...which is larger than the limit of 44GB":
```
[2019-09-11T06:19:52,728][WARN ][o.e.a.b.TransportShardBulkAction] [es002.tab.com] [[query_performance_2019.09.10][0]] failed to perform indices:data/write/bulk[s] on replica [query_performance_2019.09.10][0], node[Uc_gz1L9RlKzJrDIuDDPNg], [R], s[STARTED], a[id=ws-0IaCuRQWhSTdq7hVsvQ]
org.elasticsearch.transport.RemoteTransportException: [es003.tab.com][1.2.3.4:9300][indices:data/write/bulk[s][r]]
Caused by: org.elasticsearch.common.breaker.CircuitBreakingException: [parent] Data too large, data for [<transport_request>] would be [47571970700/44.3gb], which is larger than the limit of [47173546803/43.9gb], real usage: [47548727296/44.2gb], new bytes reserved: [23243404/22.1mb], usages [request=0/0b, fielddata=1005/1005b, in_flight_requests=23243404/22.1mb, accounting=8966251/8.5mb]
```
breakers stats:
```
      "breakers": {
        "request": {
          "limit_size_in_bytes": 29793819033,
          "limit_size": "27.7gb",
          "estimated_size_in_bytes": 0,
          "estimated_size": "0b",
          "overhead": 1,
          "tripped": 0
        }```

Zing conf:
```[root@es-data002 ~]# zing-ps --acct -h

 System Zing Memory reserved at configuration (reserve-at-config)
Fund breakdown0
                              NAME    BALANCE    MAXIMUM  COMMITTED
       fund[0]:          Committed     6144 M       56 G       56 G
       fund[1]:          Overdraft     4096 M     4096 M     4096 M
       fund[3]:    PausePrevention     4096 M     4096 M     4096 M

Found 1 process
USER         PID PROCESS
elasticsearch 170166 /opt/zing/zing-jdk11/bin/java
                                  NAME    BALANCE  ALLOCATED    MAXIMUM  FND  ODFND
    account[0]:                default      947 M     4728 K      952 M    0      1
    account[2]:                   heap       42 G     7598 M          -    0      1
    account[3]:       pause_prevention         0          0           -    3      3```

JVM options:
```[root@es002 ~]# cat /etc/elasticsearch/dba/jvm.options
-Dfile.encoding=UTF-8
-Dio.netty.noKeySetOptimization=true
-Dio.netty.noUnsafe=true
-Dio.netty.recycler.maxCapacityPerThread=0
-Djava.awt.headless=true
-Djna.nosys=true
-Dlog4j.shutdownHookEnabled=false
-Dlog4j2.disable.jmx=true
-XX:+AlwaysPreTouch
-XX:+HeapDumpOnOutOfMemoryError
-XX:+UseCMSInitiatingOccupancyOnly
-XX:-OmitStackTraceInFastThrow
-XX:CMSInitiatingOccupancyFraction=75
-Xms50g
-Xmx50g
-Xss1m
-server```

@danielmitterdorfer @dakrone  do you have any idea how to debug the issue?

**Comment by henningandersen** [comment]
Notice that the [question on zing was dealt with on discuss](https://discuss.elastic.co/t/g1gc-cause-circuitbreakingexception-parent-data-too-large-on-7-1-1/187272/37?u=henningandersen).

**Pull request merged**
